armory/art_experimental/attacks/snr_pgd.py

```diff
@@ -0,0 +1,147 @@
+from art.attacks.evasion import ProjectedGradientDescentPyTorch
+import numpy as np
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class SNR_PGD(ProjectedGradientDescentPyTorch):
+    """
+    Applies L2 PGD to signal based on an SNR bound defined by norm 'snr' or 'snr_db'.
+        This is a *lower* bound on allowable SNR (as opposed to L2 upper bound)
+
+    norm - snr or snr_db
+    eps - the lower bound on allowable SNR and SNR_DB
+    eps_step - float value in (0, 2] that is the ratio of max L2 distance per step
+        A value of 2 is the diameter of the epsilon ball.
+        NOTE: this is different from original PGD because SNR is not additive
+
+    SNR measures the original signal power to adversarial perturbation power
+        If there is no perturbation, SNR is np.inf; all other values are finite.
+        If the perturbation is unbounded, SNR is 0
+    SNR_DB = 10 * log10(SNR)
+
+    If SNR is set to 0 or SNR_DB to -inf, no limit is provided to PGD
+        If SNR or SNR_DB is set to inf, no perturbation is performed
+    """
+
+    def __init__(
+        self, estimator, norm="snr", eps=10, eps_step=0.5, batch_size=1, **kwargs
+    ):
+        if batch_size != 1:
+            raise NotImplementedError("only batch_size 1 supported")
+        super().__init__(estimator, norm=2, batch_size=1, **kwargs)
+
+        # Map to SNR domain
+        eps = float(eps)
+        if norm == "snr":
+            snr = 10 ** (eps / 10)
+        elif norm == "snr_db":
+            snr = eps
+        else:
+            raise ValueError(f"norm must be 'snr' (default) or 'snr_db', not {norm}")
+
+        if snr < 0:
+            raise ValueError(f"snr must be nonnegative, not {snr}")
+        elif snr == 0:
+            self.snr_sqrt_reciprocal = np.inf
+        elif snr == np.inf:
+            self.snr_sqrt_reciprocal = 0
+        else:
+            self.snr_sqrt_reciprocal = 1 / np.sqrt(snr)
+
+        eps_step = float(eps_step)
+
+        if not (0 < eps_step <= 2):
+            raise ValueError(f"eps_step must be in (0, 2], not {eps_step}")
+        self.step_fraction = eps_step
+
+    def generate(self, x, y=None, **kwargs):
+        x_l2 = np.linalg.norm(x, ord=2)
+        if x_l2 == 0:
+            logger.warning("Input all 0. Not making any change.")
+            return x
+        elif self.snr_sqrt_reciprocal == 0:
+            return x
+
+        eps = x_l2 * self.snr_sqrt_reciprocal
```

Comment:
I think the l2 `eps` bound should be the ratio of the l2 norm of x to the desired snr (i.e. `x_l2 / snr` without the `sqrt`)